src/notebooks/controllers/controllerRegistration.ts

```diff
@@ -82,26 +82,30 @@ export class ControllerRegistration implements IControllerRegistration {
         this.serverUriStorage.onDidRemoveUris(this.onDidRemoveUris, this, this.disposables);
     }
     batchAdd(metadatas: KernelConnectionMetadata[], types: ('jupyter-notebook' | 'interactive')[]) {
-        const added: IVSCodeNotebookController[] = [];
+        const addedList: IVSCodeNotebookController[] = [];
         metadatas.forEach((metadata) => {
-            added.push(...this.addImpl(metadata, types, false));
+            const { added } = this.addImpl(metadata, types, false);
+            addedList.push(...added);
         });
 
-        this.changeEmitter.fire({ added, removed: [] });
-        return added;
+        if (addedList.length) {
+            this.changeEmitter.fire({ added: addedList, removed: [] });
+        }
     }
-    add(
+    addOrUpdate(
         metadata: KernelConnectionMetadata,
         types: ('jupyter-notebook' | 'interactive')[]
     ): IVSCodeNotebookController[] {
-        return this.addImpl(metadata, types, true);
+        const { added, existing } = this.addImpl(metadata, types, true);
+        return added.concat(existing);
     }
     addImpl(
         metadata: KernelConnectionMetadata,
         types: ('jupyter-notebook' | 'interactive')[],
         triggerChangeEvent: boolean
-    ): IVSCodeNotebookController[] {
-        let results: IVSCodeNotebookController[] = [];
+    ): { added: IVSCodeNotebookController[]; existing: IVSCodeNotebookController[] } {
+        const added: IVSCodeNotebookController[] = [];
+        const existing: IVSCodeNotebookController[] = [];
```

Comment:
We were triggering change events even when we're updating existing items
This had other side effects, the change event only supports `added, removed`, not `updated`